tests/simulationbase.py

```diff
@@ -173,55 +171,114 @@ def _simulate(self, task_params: dict, tmpdir, expected_frames: list):
 
             task_id = 'test_task_id123'
-            req_task_work_dir = req_work_dir / task_id
-            req_task_work_dir.mkdir()
-            req_task_resources_dir = req_task_work_dir / constants.RESOURCES_DIR
-            req_task_resources_dir.mkdir()
-            req_task_net_resources_dir = \
-                req_task_work_dir / constants.NETWORK_RESOURCES_DIR
-            req_task_net_resources_dir.mkdir()
-            req_task_results_dir = req_task_work_dir / constants.RESULTS_DIR
-            req_task_results_dir.mkdir()
-            req_task_net_results_dir = \
-                req_task_work_dir / constants.NETWORK_RESULTS_DIR
-            req_task_net_results_dir.mkdir()
-
-            prov_task_work_dir = prov_work_dir / task_id
-            prov_task_work_dir.mkdir()
-            prov_task_net_resources_dir = \
-                prov_task_work_dir / constants.NETWORK_RESOURCES_DIR
-            prov_task_net_resources_dir.mkdir()
+            req_task_work_dir, req_task_resources_dir, \
+                req_task_net_resources_dir, req_task_results_dir, \
+                req_task_net_results_dir = \
+                self._make_req_dirs(req_work_dir, task_id)
+
+            prov_task_work_dir, prov_task_net_resources_dir = \
+                self._make_prov_dirs(prov_work_dir, task_id)
 
             self._put_cube_to_resources(req_task_resources_dir)
 
             self._create_task(requestor, task_id, task_params)
 
             for _ in range(task_params['subtasks_count']):
 
-                subtask_id, subtask_params = \
-                    self._get_next_subtask(requestor, task_id)
-                assert subtask_params['resources'] == [0]
-
-                self._copy_resources_from_requestor(
+                self._do_subtask(
+                    task_id,
+                    requestor,
+                    provider,
                     req_task_net_resources_dir,
+                    req_task_net_results_dir,
                     prov_task_net_resources_dir,
-                    subtask_params,
+                    prov_task_work_dir
                 )
-                prov_subtask_work_dir = prov_task_work_dir / subtask_id
-                prov_subtask_work_dir.mkdir()
-                self._compute(provider, task_id, subtask_id, subtask_params)
-                self._copy_result_from_provider(
-                    prov_subtask_work_dir,
+
+            for frame in expected_frames:
+                filename = f'result{frame:04d}.{task_params["format"]}'
+                result_file = req_task_results_dir / filename
+                assert result_file.exists()
+        finally:
+            self._close_server(requestor_server)
+            self._close_server(provider_server)
+
+
+    def _simulate_restart(self, task_params: dict, tmpdir,
+                          expected_frames: list):
+
+        skipping_subtasks = 1
+
+        tmpdir = Path(tmpdir)
+        req_work_dir = tmpdir / 'requestor'
```

Comment:
It'd be greater to deduplicate these lines below. Could be just setting everything up in the setup method.

Reply:
Done, for now i made _setup_*() methods for requestor and provider